### ISSUE TITLE
Failed to send/receive btrfs subvolume

### ISSUE THREAD

**Issue by huegelc** [digint/btrbk#241]
Hi,

I'm using following config to backup/snapshot lxd containers and send them via ssh but this fails with:

`ERROR: Failed to send/receive btrfs subvolume: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont01.20180710T1007  -> example.com:/nas-backup` 

`Backup Summary (btrbk command line client, version 0.26.0)`

```
# test config
timestamp_format long
transaction_syslog local7
snapshot_preserve_min 24h
target_preserve_min 24h
ssh_identity /root/.ssh/id_rsa
ssh_port 2006
volume /var/snap/lxd/common/lxd/storage-pools/default/containers

subvolume bckp-cont05
  stream_compress lzo
  stream_compress_level 3
  btrfs_commit_delete each
    ssh_user root
    target send-receive ssh://example.com/nas-backup
```
The other way around (from example.com -> local) works. 
Any hints are appreciated.

**Comment by digint** [comment]
This looks like a common / sane config. Maybe you are missing the `lzop` command used by `stream_compress lzo`?

Without debug log I can't help much, please `btrbk -l debug run` and paste the relevant parts.

**Comment by huegelc** [comment]
Thx for your reply.

```
btrbk -l debug run -c /etc/btrbk/btrbk-nas.conf
btrbk command line client, version 0.26.0  (Tue Jul 10 14:25:10 2018)
Using configuration: /etc/btrbk/btrbk-nas.conf
Initializing volume section: /var/snap/lxd/common/lxd/storage-pools/default/containers
### btrfs subvolume show /var/snap/lxd/common/lxd/storage-pools/default/containers
Command execution successful
Parsed 11 subvolume detail items: /var/snap/lxd/common/lxd/storage-pools/default/containers
### btrfs subvolume list -a -c -u -q -R /var/snap/lxd/common/lxd/storage-pools/default/containers
Command execution successful
Parsed 291 total subvolumes for filesystem at: /var/snap/lxd/common/lxd/storage-pools/default/containers
### btrfs subvolume list -a -r /var/snap/lxd/common/lxd/storage-pools/default/containers
Command execution successful
Parsed 263 readonly subvolumes for filesystem at: /var/snap/lxd/common/lxd/storage-pools/default/containers
Initializing subvolume section: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05
Found "/var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05" (id=30719) in btrfs subtree of: /var/snap/lxd/common/lxd/storage-pools/default/containers
Initializing target section: example.com:/nas-backup
### ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume show /nas-backup'
Command execution successful
Parsed 11 subvolume detail items: example.com:/nas-backup
### ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume list -a -c -u -q -R /nas-backup'
Command execution successful
Parsed 258 total subvolumes for filesystem at: example.com:/nas-backup
### ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume list -a -r /nas-backup'
Command execution successful
Parsed 232 readonly subvolumes for filesystem at: example.com:/nas-backup
Opening syslog
Syslog enabled
Snapshot creation enabled (snapshot_create=always)
Creating subvolume snapshot for: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05
[snapshot] source: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05
[snapshot] target: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425
### btrfs subvolume snapshot -r /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05 /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425
Command execution successful
Found 1 snapshot children of "/var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05" in: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.*
Checking for missing backups of subvolume "/var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05" in "example.com:/nas-backup/"
Adding backup candidate: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425
Checking schedule for backup candidates
Schedule: preserving all within 24 hours
Scheduler reference time: 2018-07-10T14:25:10+0200
Preserving 1/1 items
Creating subvolume backup (send-receive) for: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425
No common subvolumes of "/var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425" found in src="/var/snap/lxd/common/lxd/storage-pools/default/containers/", target="example.com:/nas-backup/"
No common parent subvolume present, creating full backup...
[send/receive] source: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425
[send/receive] target: example.com:/nas-backup/bckp-cont05.20180710T1425
### btrfs send /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425 | lzop -c -3 | ssh -p 2006 -i /root/.ssh/id_rsa -o compression=no root@example.com 'lzop -d -c | btrfs receive /nas-backup/'
Command execution successful
[send/receive] checking target metadata: example.com:/nas-backup/bckp-cont05.20180710T1425
### ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume show /nas-backup/bckp-cont05.20180710T1425'
Command execution failed (exitcode=1): `ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume show /nas-backup/bckp-cont05.20180710T1425'`
ERROR: Failed to send/receive btrfs subvolume: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425  -> example.com:/nas-backup
send/received failed, deleting (possibly present and garbled) received subvolume: example.com:/nas-backup/bckp-cont05.20180710T1425
[delete] options: commit-after
[delete] target: example.com:/nas-backup/bckp-cont05.20180710T1425
### ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume delete --commit-after /nas-backup/bckp-cont05.20180710T1425'
Command execution failed (exitcode=1): `ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume delete --commit-after /nas-backup/bckp-cont05.20180710T1425'`
ERROR: Failed to delete subvolume "example.com:/nas-backup/bckp-cont05.20180710T1425": ERROR: cannot access subvolume /nas-backup/bckp-cont05.20180710T1425: No such file or directory
WARNING: Deletion of partially received (garbled) subvolume failed, assuming clean environment: example.com:/nas-backup/bckp-cont05.20180710T1425
ERROR: Error while resuming backups, aborting
Created 0/1 missing backups
Found 1 snapshot children of "/var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05" in: /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.*
WARNING: Skipping cleanup of snapshots for subvolume "/var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05", as at least one target aborted earlier
Completed within: 4s  (Tue Jul 10 14:25:14 2018)
Closing syslog
--------------------------------------------------------------------------------
Backup Summary (btrbk command line client, version 0.26.0)

    Date:   Tue Jul 10 14:25:10 2018
    Config: /etc/btrbk/btrbk-nas.conf

Legend:
    ===  up-to-date subvolume (source snapshot)
    +++  created subvolume (source snapshot)
    ---  deleted subvolume
    ***  received subvolume (non-incremental)
    >>>  received subvolume (incremental)
--------------------------------------------------------------------------------
/var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05
+++ /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425
!!! example.com:/nas-backup/bckp-cont05.20180710T1425
!!! Target "example.com:/nas-backup" aborted: Failed to send/receive subvolume

NOTE: Some errors occurred, which may result in missing backups!
Please check warning and error messages above.
```

**Comment by digint** [comment]
Looks like the backup gets transferred:
```
### btrfs send /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425 | lzop -c -3 | ssh -p 2006 -i /root/.ssh/id_rsa -o compression=no root@example.com 'lzop -d -c | btrfs receive /nas-backup/'
Command execution successful
```

But then checking the target subvolume fails:

```
### ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume show /nas-backup/bckp-cont05.20180710T1425'
Command execution failed (exitcode=1): `ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume show /nas-backup/bckp-cont05.20180710T1425'`
```

I think old versions of btrfs-progs behave like this.
What version of btrfs-progs do you use (`btrfs --version` on both source and example.com)?
Is the `lzop` command available on example.com?
What happens if you run these commands in the shell? :
```
btrfs send /var/snap/lxd/common/lxd/storage-pools/default/containers/bckp-cont05.20180710T1425 | lzop -c -3 | ssh -p 2006 -i /root/.ssh/id_rsa -o compression=no root@example.com 'lzop -d -c | btrfs receive /nas-backup/'
ssh -p 2006 -i /root/.ssh/id_rsa root@example.com 'btrfs subvolume show /nas-backup/bckp-cont05.20180710T1425'
```

(tip: if you run `btrbk -l trace`, you will also see the output of the commands issued by btrbk, and possibly some helpful error messages)

**Comment by huegelc** [comment]
You´re right `lzo` package was missing on the target. 
Would it be a good idea to make btrfs handle this kind of errors more specific? Im not a develeoper so Im just blind guessing here.

**Comment by digint** [comment]
> Would it be a good idea to make btrbk handle this kind of errors more specific? Im not a develeoper so Im just blind guessing here.

Actually this was a bug in the error handling. Btrbk should have printed `bash: lzop: command not found` (stderr message from the underlying shell)

fixed in: de57efff82bda540cea378029e0233b13e7715a5
